Docs for communication API and updates for provider plugin type - MDL-79961

safatshahin opened the pull request:
This doc is related to Communication API and provider plugin type.

To track the progress of this task, I used MDL-79961 in our current sprint.

safatshahin committed:
[docs] Communication subsystem docs for MDL-79961

safatshahin committed:
[docs] Communication plugintype doc update for MDL-79961

sarjona commented:
@safatshahin is this pull-request blocked by https://tracker.moodle.org/browse/MDL-78551 (so should wait for this issue to be integrated before merging it)?

Apart from that, it would be great if you could fix the failures ;-)

safatshahin commented:
Hi @sarjona 
I have removed the dependency of MDL-78551 and also moved the docs specifically related to that tracker to a different branch, I will push that later while 78551 is in integration, and we can move this forward as we don't have comm API docs and the plugin doc is not up to date.
I have resolved the CI errors, please give it another go and you can remove the blocked-by-tracker label.
Cheers!

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+:::info
+
+This is an experimental feature. To use this feature, you must enable the experimental feature flag for Communication in the site administration.
```
I think it would be better to give the exact name of this experimental setting, to help admins find it, something like Enable communication providers (enablecommunicationsubsystem).

safatshahin replied:
Done, added the steps for better understanding.

sarjona commented on docs/apis/subsystems/communication/index.md:
There is a typo in the commit message: Communicaiton

sarjona commented on docs/apis/subsystems/communication/index.md:
I think the changes on this pull request should be backported to the versioned_docs/version-4.3 branch (or at least backport the ones that apply to Moodle 4.3).

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+### Loading a communication instance
+
+To load a communication instance, you must use the communication\api::load_by_instance() method. This method will return a communication instance which can be
```
```suggestion
To load a communication instance, you must use the `communication\api::load_by_instance()` method. This method will return a communication instance which can be
```

safatshahin replied:
done

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+To load a communication instance, you must use the communication\api::load_by_instance() method. This method will return a communication instance which can be
+used to interact with the communication API and its related methods. The below example shows how to load a communication instance for a course, where the
+context is the actual course context object, component is the component name (core_course), instance type is the custom string which can change according to
```
```suggestion
context is the actual course context object, component is the component name (`core_course`), instance type is the custom string which can change according to
```

safatshahin replied:
done

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+used to interact with the communication API and its related methods. The below example shows how to load a communication instance for a course, where the
+context is the actual course context object, component is the component name (core_course), instance type is the custom string which can change according to
+the usage and instance id is the course id. The constructor of the public api is private and hence the only way to load an instance is through
+the load_by_instance() method. The provider is the name of the provider plugin which is used to load the instance. It's optional, if not provided, it will load
```
```suggestion
the `load_by_instance()` method. The provider is the name of the provider plugin which is used to load the instance. It's optional, if not provided, it will load
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Create and configure a room
+
+$communication->create_and_configure_room() method is used to add an ad-hoc to create a room in the communication provider and configure it with the required settings.
```
```suggestion
`$communication->create_and_configure_room()` method is used to add an ad-hoc to create a room in the communication provider and configure it with the required settings.
```

sarjona commented on docs/apis/subsystems/communication/index.md:
````diff
@@ -0,0 +1,244 @@
+Please refer to [Communication plugin](../../plugintypes/communication/index.md) documentation for more details.
+
+```php
+// first initialize the communication instance, provided the context is already initialized and the selected provider is available though a form or other means.
````
```suggestion
// First initialize the communication instance, provided the context is already initialized and the selected provider is available
// though a form or other means.
```

sarjona replied:
The given examples should follow the coding style rules (because they will be taken as a good practices to follow) ;-)

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+            provider: 'communication_matrix',
+        );
+
+// Now call the create_and_configure_room() method to add an ad-hoc to create a room in the communication provider and configure it with the required settings.
```
```suggestion
// Now call the create_and_configure_room() method to add an ad-hoc to create a room in the communication provider and configure it
// with the required settings.
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Update a room and its associated information
+
+$communication->update_room() method is used to add an ad-hoc to update a room in the communication provider and configure it with the required settings.
```
```suggestion
`$communication->update_room()` method is used to add an ad-hoc to update a room in the communication provider and configure it with the required settings.
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+            provider: 'communication_matrix',
+        );
+
+// Now update the room with the required settings. The instance will be used by dynamic fields feature of the plugin to allow the plugin to get the required information from the instance.
```
This probably also needs to be reviewed (the comment should be split in different lines).

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Delete a room
+
+$communication->delete_room() method is used to add an ad-hoc to delete a room in the communication provider. The associated task for this also removes all the members from
```
```suggestion
`$communication->delete_room()` method is used to add an ad-hoc to delete a room in the communication provider. The associated task for this also removes all the members from
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Add members to a room
+
+$communication->add_members_to_room() method is used to add an ad-hoc to add members to a room in the communication provider. The user id of each user to be added to the provider room is also stored in
```
```suggestion
`$communication->add_members_to_room()` method is used to add an ad-hoc to add members to a room in the communication provider. The user id of each user to be added to the provider room is also stored in
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Update the membership of a room
+
+$communication->update_room_membership() method is used to add an ad-hoc to update the membership of a room in the communication provider. The user mapping for each of the users are also reset to allow task to update the users from the task.
```
```suggestion
`$communication->update_room_membership()` method is used to add an ad-hoc to update the membership of a room in the communication provider. The user mapping for each of the users is also reset to allow task to update the users from the task.
```

sarjona commented on docs/apis/subsystems/communication/index.md:
````diff
@@ -0,0 +1,244 @@
+    )
+```
+
+This method accepts the same parameters as the add_members_to_room() method and the usage is also the same.
````
```suggestion
This method accepts the same parameters as the `add_members_to_room()` method and the usage is also the same.
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Remove members from a room
+
+$communication->remove_members_from_room() method is used to add an ad-hoc to remove members from a room in the communication provider. The users are flagged as deleted in the communication_user table to allow the task to remove the users from the provider room.
```
```suggestion
`$communication->remove_members_from_room()` method is used to add an ad-hoc to remove members from a room in the communication provider. The users are flagged as deleted in the `communication_user` table to allow the task to remove the users from the provider room.
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+#### Delete a room
+
+$communication->delete_room() method is used to add an ad-hoc to delete a room in the communication provider. The associated task for this also removes all the members from
+the room before deleting the room. This is destructive and might remove all the messages and other data associated with the room. Usage of this should be done with caution.
```
```suggestion
the room before deleting the room.

:::danger

This is destructive and might remove all the messages and other data associated with the room. Usage of this should be done with caution.

:::

```

sarjona commented on docs/apis/subsystems/communication/index.md:
````diff
@@ -0,0 +1,244 @@
+    )
+```
+
+This method accepts the same parameters as the add_members_to_room() method and the usage is also the same.
````
```suggestion
This method accepts the same parameters as the `add_members_to_room()` method and the usage is also the same.
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Show the communication room creation status notification
+
+$communication->show_communication_room_status_notification() is a special method to have a UI element to show the status of the room creation. If the room is ready, it will
```
```suggestion
`$communication->show_communication_room_status_notification()` is a special method to have a UI element to show the status of the room creation. If the room is ready, it will
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+:::
+
+### Loading a communication instance
```
```suggestion
## Loading a communication instance
```

sarjona replied:
The first level should be ## so all the headings in this file should be reviewed too.

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Communication instance setup using form/configuration page
+
+The communication API allows to have settings to configure the basic information like room name, selection of provider etc. $communication->form_definition() method
```
```suggestion
The communication API allows settings to configure the basic information like room name, selection of provider etc. `$communication->form_definition()` method
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+#### Communication instance setup using form/configuration page
+
+The communication API allows to have settings to configure the basic information like room name, selection of provider etc. $communication->form_definition() method
+is used to get the form definition for the settings form. If the form elements are used in another form to include the communication form elements, form_definition() method
```
```suggestion
is used to get the form definition for the settings form. If the form elements are used in another form to include the communication form elements, `form_definition()` method
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+
+The communication API allows to have settings to configure the basic information like room name, selection of provider etc. $communication->form_definition() method
+is used to get the form definition for the settings form. If the form elements are used in another form to include the communication form elements, form_definition() method
+will be useful. There is a configuration page (communication/configure.php) which can be used to configure the communication instance. Course currently uses this page to set up
```
```suggestion
will be useful. There is a configuration page (`communication/configure.php`) which can be used to configure the communication instance. Course currently uses this page to set up
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+is used to get the form definition for the settings form. If the form elements are used in another form to include the communication form elements, form_definition() method
+will be useful. There is a configuration page (communication/configure.php) which can be used to configure the communication instance. Course currently uses this page to set up
+communication and its associated information.
+It will also be required to use set_data() method in order to set the data to the form elements while saving the data.
```
```suggestion
It will also be required to use `set_data()` method in order to set the data to the form elements while saving the data.
```

sarjona commented on docs/apis/subsystems/communication/index.md:
```diff
@@ -0,0 +1,244 @@
+### Building a communication plugin
+
+Communication API also provides a bunch of interfaces for a communication plugin to consume. Every plugin should implement these interfaces according to the features they
+support. Please refer to [Communication plugin](../../plugintypes/communication/index.md) documentation for more details about building a plugin.
```
```suggestion
support. 


:::info

Please refer to [Communication plugin](../../plugintypes/communication/index.md) documentation for more details about building a plugin.

:::
```

sarjona commented on docs/apis/plugintypes/communication/index.md:
```diff
@@ -80,8 +72,9 @@ There are a number of key files within the plugin, described below.
-Each plugin must implement this class and should have the exact class name. This part is important for the core communication api.
-The core communication api will pick the features and actions from this class. This class should implement the interfaces it supports.
+Each plugin must implement this class and should have the exact class name. The core communication api will pick the features and actions from this class. There is no strict rule
+on which interfaces should be implemented, plugins can choose which features they support and implement accordingly. Exception is the communication_provider interface which
```
```suggestion
on which interfaces should be implemented, plugins can choose which features they support and implement accordingly. Exception is the `communication_provider` interface which
```

sarjona commented on docs/apis/plugintypes/communication/index.md:
```diff
@@ -100,24 +93,26 @@ class communication_feature implements
-This is the base user provider interface. This interface should be used to declare the support for the for user provider creation and management.
-Every provider plugin must implement this interface if user creation is required. This interface will have the following methods.
+This is the user provider interface. This interface should be used to declare the support for the for user creation for a provider. For example, Matrix allows creation of users
+via API and the communication_matrix plugin can support the creation of users in Matrix, in that case, communication_matrix plugin should implement this interface. Some APIs might
```
```suggestion
via API and the `communication_matrix` plugin can support the creation of users in Matrix, in that case, `communication_matrix` plugin should implement this interface. Some APIs might
```

sarjona commented on docs/apis/plugintypes/communication/index.md:
```diff
@@ -134,20 +129,36 @@ value should be returned to indicate if the room is deleted or something went wr
 
+### synchronise_provider
+
+Communication API has a scheduled task core_communication\task\synchronise_providers_task which will synchronise the data from the communication provider to the current Moodle
```
```suggestion
Communication API has a scheduled task `core_communication\task\synchronise_providers_task` which will synchronise the data from the communication provider to the current Moodle
```

sarjona reviewed (changes requested):
Hi @safatshahin!
Thanks for clarifying and updating the dependencies :-)
The docs look good. However, as you'll see, I've raised a few points that should be reviewed ;-)
If you have any question, let me know.

safatshahin committed:
[docs] Communication plugintype 4.3 updates for MDL-79961

safatshahin committed:
[docs] Fork Version 4.3 docs for Communication susbystem

safatshahin commented:
Hi @sarjona 
Thank you for your comments. I have addressed all of them, added the 4.3 forked version for the subsystem, and updated the 4.3 docs for plugin type. The current doc will not need any changes for 4.3 as the changes related to 4.4 are done in a different branch.

I have 4 commits to make changes in 4 different files which should make it easier to review.

Please give it another go and let me know of any other changes.

Cheers!

safatshahin commented:
Hi @sarjona 
The tracker is closed so that the review can continue in the GH issue.
Please let me know if you want any other changes.
Cheers!

sarjona reviewed (approved):
Hi @safatshahin!
Thanks a lot for reviewing the raised points. The docs look good so I'm happy to approve this pull-request :-)
